<a href="https://colab.research.google.com/github/cheungngo/MScGenomics/blob/master/5030/HW1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# GNBF5030 Homework 1

### CHEUNG NGO

*** This hw is done in the google colab environment, so that "!" will appear before each command

Please save your homework in **PDF format** and name it as **LastnameFirstname-hw1.pdf** (without any space). Submit it to the Blackboard on or before **Friday 11 February**.

You will need to log into the MSc student server to work on these questions. Data files are available under `/hdd1/shared/ecoli/` and `/hdd1/shared/bdgp6/data/genomes`. For each question, list your commands and output (use `head` to display the first few lines if the output is too long). For example,

$ wc -l SRR098026.fastq

996 SRR098026.fastq

## Question 1

The four nucleotides that appear in DNA are abbreviated A, C, T and G. Unknown nucleotides are represented with the letter N. An N appearing in a sequencing file represents a position where the sequencing machine was not able to confidently determine the nucleotide in that position. List your commands and output, if any, to answer the following questions.


### 1. Create a directory ~/5030/hw1 for Homework 1 and copy the ecoli foder under /hdd1/shared/ to ~/5030/hw1.


In [ ]:
!mkdir -p 5030/hw1

In [ ]:
!cp "/content/drive/MyDrive/Colab Notebooks/5030/ecoli" 5030/hw1/ -r

In [ ]:
# check if it is copied
%cd 5030/hw1
!ls

/content/5030/hw1
ecoli


### 2. How many sequences in `ecoli/untrimmed_fastq/SRR098026.fastq` contain the `TTCNNNNN`? Use `grep` and `wc -l` in a single command to find the number.


In [ ]:
!grep TTCNNNNN ecoli/untrimmed_fastq/SRR098026.fastq | wc -l
# -l is for the line count

4


In [ ]:
# visualizing the data
!grep TTCNNNNN ecoli/untrimmed_fastq/SRR098026.fastq | head

NNNNNNNNNNNTCGCTTCNNNNNNNNNNNNNCNNN
ANNNNNNNNNCTTTTTTTTCNNNNNNNNNATTCTN
CNNNNNNNNNCTCCCCTTTCNNNNNNNNNTAATTN
ANNNNNNNACAACGCTGTTTCNNNNNNNNGTTCGC


### 3. For both `fastq` files under `ecoli/untrimmed_fastq`, search for sequences with 8 consecutive unknown nucleotides (Ns). Have your search report each sequence with header line (or identifier) and redirect the results to an output file `bad_reads.txt`. Your search should be done in one single command. To capture the header line, use `-B` in `grep` command. Your search should also return a specific number of lines before each match. *Note: It can be really useful to do this type of searching to get a feel for the quality of your sequencing results. However, in you research you will most likely use a bioinformatics tool that has a built-in program for filtering out low-quality reads. We will learn such tools soon.*


In [ ]:
# find the file names under ecoli/untrimmed_fastq
!ls ecoli/untrimmed_fastq

SRR097977.fastq  SRR098026.fastq


In [ ]:
# checking on the number of bad sequences
!grep NNNNNNNN ecoli/untrimmed_fastq/SRR097977.fastq | wc -l

0


In [ ]:
# checking on the number of bad sequences
!grep NNNNNNNN ecoli/untrimmed_fastq/SRR098026.fastq | wc -l

223


In [ ]:
# doing the operation
!grep -B 1 NNNNNNNN ecoli/untrimmed_fastq/SRR098026.fastq >> bad_reads.txt

In [ ]:
# visualizing the bad_reads.txt
!head bad_reads.txt

@SRR098026.1 HWUSI-EAS1599_1:2:1:0:968 length=35
NNNNNNNNNNNNNNNNCNNNNNNNNNNNNNNNNNN
--
@SRR098026.2 HWUSI-EAS1599_1:2:1:0:312 length=35
NNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNNNN
--
@SRR098026.3 HWUSI-EAS1599_1:2:1:0:570 length=35
NNNNNNNNNNNNNNNNANNNNNNNNNNNNNNNNNN
--
@SRR098026.4 HWUSI-EAS1599_1:2:1:0:758 length=35


### 4. From your home directory, create a new folder `q1` under `~/5030/hw1/`. Then go to this new directory and copy the output file from the previous question to this directory, using **relative path**. At the end, display information of this output file in **long format** using `ls` command and show the number of lines using `wc` command. List all your commands starting from `cd`, as well as results from `ls` and `wc`.


In [ ]:
!echo $PWD

/content/5030/hw1


In [ ]:
# setting the relative path
%env FILEPATH=/content/5030/hw1/bad_reads.txt

env: FILEPATH=/content/5030/hw1/bad_reads.txt


In [ ]:
# make the new directory
!mkdir q1

In [ ]:
# copy using relative path
!cp $FILEPATH q1/bad_reads.txt

In [ ]:
# cd to the q1 folder and see if the file has been copied
%cd q1
!ls

/content/5030/hw1/q1
bad_reads.txt


In [ ]:
# display information of this output file in long format using ls command
!ls -l bad_reads.txt

-rw-r--r-- 1 root root 20060 Jan 26 09:47 bad_reads.txt


In [ ]:
# number of lines
!wc -l bad_reads.txt

668 bad_reads.txt


## Question 2

Suppose that the SRA metadata file `ecoli/sra_metadata/SraRunTable.txt` contains a lot of information about the samples that we submitted for sequencing. The samples were a mix of single- and paired-end libraries. Now we want to know how many samples of each library type are recorded in our table. Follow steps below to explore the table and perform your task. First of all, copy the original file `SraRunTable.txt` to your own directory `~/5030/hw1/q2`.


In [ ]:
!echo $PWD

/content/5030/hw1


In [ ]:
# make new directory
!mkdir q2

In [ ]:
# copy the file
!cp ecoli/sra_metadata/SraRunTable.txt q2/SraRunTable.txt

In [ ]:
# check if it is copied
%cd q2
!ls

/content/5030/hw1/q2
SraRunTable.txt


### 1. The column headers are in the first row of the meta table. Use `head -n` command to check the first line of the table and record the number of column for `LibraryLayout_s`.


In [ ]:
!head -n 1 SraRunTable.txt | awk -v b="LibraryLayout_s" '{for (i=1;i<=NF;i++) { if ($i == b) { print i } }}'

3


### 2. Use `cut` and `head` to view the first 12 lines of the `LibraryLayout_s` column, in one single command.


In [ ]:
!head -n 12 SraRunTable.txt | cut -f3

LibraryLayout_s
SINGLE
SINGLE
SINGLE
SINGLE
SINGLE
SINGLE
SINGLE
SINGLE
PAIRED
SINGLE
SINGLE


### 3. Use `cut`, `grep -v` to extract the column of `LibraryLayout_s` without the header and redirect the output to `library_layout.txt`. Then use `wc` command to double check that your output file contains the right number of lines. *Help: The `grep -v` command can help you remove the header.]*


In [ ]:
# extract the column without the header
!grep -v "LibraryLayout_s" SraRunTable.txt | cut -f3 >> library_layout.txt

In [ ]:
# check on the number of lines in the two files
!wc -l SraRunTable.txt
!wc -l library_layout.txt

38 SraRunTable.txt
37 library_layout.txt


### 4. Count the number of `PAIRED` and `SINGLE` in `library_layout.txt` using `sort` and `uniq -c`. *Help: The `sort` command will sort the lines of a text file and the `uniq` command will filter out repeated neighboring lines in a file. You might expect `uniq` to extract all of the unique lines in a file. But it isn't.*


In [ ]:
! sort library_layout.txt | uniq -c

      2 PAIRED
     35 SINGLE


### 5. Combine commands in questions 3 and 4 into a single one to get number of samples of each library type (with piping), so that no intermediate file is needed.


In [ ]:
!grep -v "LibraryLayout_s" SraRunTable.txt | cut -f3 | sort | uniq -c

      2 PAIRED
     35 SINGLE


### 6. Similarly, can you find out the number of different sample load dates in the table? How many samples were loaded on each date? Perform this task in a single command.


In [ ]:
# get the column number first
# (could be performed by simple visualization though)
!head -n 1 SraRunTable.txt | awk -v b="LoadDate_s" '{for (i=1;i<=NF;i++) { if ($i == b) { print i } }}'

5


In [ ]:
# get the number of different sample load dates
!grep -v "LoadDate_s" SraRunTable.txt | cut -f5 | sort | uniq -c

      6 25-Jul-12
     31 29-May-14


## Question 3

Data files are available under `/hdd1/shared/bdgp6/data/genomes`.

### 1. Prepare your work space by creating directories of the following hierarchy.

    ```
    ~/5030/hw1/q3/
    ├── summary
    |   └── q3_directory_contents.txt
    ├── genome-annot
    │   └── BDGP6.Ensembl.93.gtf
    └── output
    ```
  a. Create directories of `summary`, `genome-annot` and `output`, under `~/5030/hw1/q3`.

  b. Under `genome-annot`, create a symbolic link of the `.gft` file using `ln -s` command. The original `.gtf` file is under `/hdd1/shared/bdgp6/data/genomes`.

  c. Record your directory structure by saving the output of `ls -lR ~/5030/hw1/q3` into the `summary` directory as a file named `q3_directory_contents.txt`.


In [ ]:
# making the directories

!mkdir 5030
%cd 5030
!mkdir hw1
%cd hw1
!mkdir q3
%cd q3
!mkdir summary
!mkdir genome-annot
!mkdir output
!ls

mkdir: cannot create directory ‘5030’: File exists
/content/5030
/content/5030/hw1
/content/5030/hw1/q3
genome-annot  output  summary


In [ ]:
# creating the link

%cd genome-annot
!ln -s "/content/drive/MyDrive/Colab Notebooks/5030/genomes/BDGP6.Ensembl.93.gtf"

/content/5030/hw1/q3/genome-annot


In [ ]:
%cd ..

In [ ]:
# Record your directory structure

!ls -lR > summary/q3_directory_contents.txt

In [ ]:
# looking at the directory structure

!head summary/q3_directory_contents.txt

.:
total 12
drwxr-xr-x 2 root root 4096 Feb  2 08:26 genome-annot
drwxr-xr-x 2 root root 4096 Feb  2 08:23 output
drwxr-xr-x 2 root root 4096 Feb  2 08:27 summary

./genome-annot:
total 4
lrwxrwxrwx 1 root root 72 Feb  2 08:26 BDGP6.Ensembl.93.gtf -> /content/drive/MyDrive/Colab Notebooks/5030/genomes/BDGP6.Ensembl.93.gtf



### 2. In `BDGP6.Ensembl.93.gtf`, how many different `transcript_id`s are associated with the gene `Sxl`?


In [ ]:
%cd genome-annot

/content/5030/hw1/q3/genome-annot


In [ ]:
!head BDGP6.Ensembl.93.gtf

#!genome-build BDGP6
#!genome-version BDGP6
#!genome-date 2014-07
#!genome-build-accession GCA_000001215.4
3R	FlyBase	gene	567076	2532932	.	+	.	gene_id "FBgn0267431"; gene_name "Myo81F"; gene_source "FlyBase"; gene_biotype "protein_coding";
3R	FlyBase	transcript	567076	2532932	.	+	.	gene_id "FBgn0267431"; transcript_id "FBtr0392909"; gene_name "Myo81F"; gene_source "FlyBase"; gene_biotype "protein_coding"; transcript_name "Myo81F-RB"; transcript_source "FlyBase"; transcript_biotype "protein_coding";
3R	FlyBase	exon	567076	567268	.	+	.	gene_id "FBgn0267431"; transcript_id "FBtr0392909"; exon_number "1"; gene_name "Myo81F"; gene_source "FlyBase"; gene_biotype "protein_coding"; transcript_name "Myo81F-RB"; transcript_source "FlyBase"; transcript_biotype "protein_coding"; exon_id "FBtr0392909-E1";
3R	FlyBase	exon	835376	835491	.	+	.	gene_id "FBgn0267431"; transcript_id "FBtr0392909"; exon_number "2"; gene_name "Myo81F"; gene_source "FlyBase"; gene_biotype "protein_coding"; transcript_name 

In [ ]:
!cut -f9 BDGP6.Ensembl.93.gtf | grep "Sxl" | grep "transcript_id" | wc -l

488


### 3. How many different features (column 3) in `BDGP6.Ensembl.93.gtf` file?

  a. Use `cut` and `head` to view part of the contents in column 3.

  b. In a single line of command, use `grep -v '^#'` to remove the header lines and then use `cut` and `head` to view part of contents in column 3 again. (Just print first 10 lines, no need to save any file.) What does `grep -v '^#'` mean? (`^` represents the beginning of a line)

  c. Modify command in question `b` to list all different types of features in the `.gtf` file. (Use `sort` and `uniq`; in a single line of command)

  d. Modify command in question `c` to show the number of each feature type and sort it in the order of the numbers using `sort -n` command. (Perform the task in a single line of command)


In [ ]:
# Part a

!cut -f3 BDGP6.Ensembl.93.gtf | head

#!genome-build BDGP6
#!genome-version BDGP6
#!genome-date 2014-07
#!genome-build-accession GCA_000001215.4
gene
transcript
exon
exon
CDS
start_codon


In [ ]:
# Part b

!grep -v '^#' BDGP6.Ensembl.93.gtf | cut -f3 | head
# '^#' means starting with #

gene
transcript
exon
exon
CDS
start_codon
exon
CDS
exon
CDS


In [ ]:
# Part c

!grep -v '^#' BDGP6.Ensembl.93.gtf | cut -f3 | sort | uniq

CDS
exon
five_prime_utr
gene
Selenocysteine
start_codon
three_prime_utr
transcript


In [ ]:
# Part d

!grep -v '^#' BDGP6.Ensembl.93.gtf | cut -f3 | sort | uniq -c | sort -nr

 187373 exon
 160859 CDS
  46299 five_prime_utr
  34767 transcript
  33892 three_prime_utr
  30492 start_codon
  17737 gene
      4 Selenocysteine


### 4. Create a simpler `.bed` file containing only genes, excluding those on scaffolds beginning with `211` (The first column of `.gtf` file is name of chromosome or scaffold). For details of `GTF` and `BED` file formats, see https://asia.ensembl.org/info/website/upload/gff.html and https://asia.ensembl.org/info/website/upload/bed.html.


  a. The combination of `awk` and `grep` can get you part of the way. Please try the following command and interpret it in your own words.
  
    ```
    $ cd ~/5030/hw1/q3/genome-annot
    $ awk '$3 == "gene" {print $1 "\t" $4 "\t" $5}' BDGP6.Ensembl.93.gtf \
    | grep -v ^211 | head
    ```

In [ ]:
!awk '$3 == "gene" {print $1 "\t" $4 "\t" $5}' BDGP6.Ensembl.93.gtf | grep -v ^211 | head

3R	567076	2532932
3R	722370	722621
3R	1031171	1031354
3R	1366234	1366601
3R	1865108	1866008
3R	2156916	2157206
3R	2554162	3263582
3R	2744304	2744800
3R	3322810	3354486
3R	3461351	3587054


  b. Modify the `awk` `print` command so that column 4 is the gene name and column 5 is the gene length. e.g.
  ```
  3R	4641963	4644636	"slx1";	2674
  3R	4641963	4643292	"MED31";	1330
  3R	4644649	4645655	"Dlip2";	1007
  3R	4645981	4648884	"rpk";	2904
  3R	4646433	4648242	"CR44155";	1810
  ```

In [ ]:
!awk '$3 == "gene" {print $1 "\t" $4 "\t" $5 "\t" $12 "\t" $5-$4}' BDGP6.Ensembl.93.gtf | grep -v ^211 | head

3R	567076	2532932	"Myo81F";	1965856
3R	722370	722621	"CR41571";	251
3R	1031171	1031354	"CR12798";	183
3R	1366234	1366601	"CR46123";	367
3R	1865108	1866008	"CR46122";	900
3R	2156916	2157206	"CR40182";	290
3R	2554162	3263582	"Pzl";	709420
3R	2744304	2744800	"CR45220";	496
3R	3322810	3354486	"spok";	31676
3R	3461351	3587054	"Parp";	125703


c. Continue to modify the command and use `sed` to remove the `;` sign at the end of gene_name column. (You may want to use `\;` instead of `;` in `sed` substitution)


In [ ]:
!awk '$3 == "gene" {print $1 "\t" $4 "\t" $5 "\t" $12 "\t" $5-$4}' BDGP6.Ensembl.93.gtf \
| grep -v ^211 | sed s/';'//g | head

3R	567076	2532932	"Myo81F"	1965856
3R	722370	722621	"CR41571"	251
3R	1031171	1031354	"CR12798"	183
3R	1366234	1366601	"CR46123"	367
3R	1865108	1866008	"CR46122"	900
3R	2156916	2157206	"CR40182"	290
3R	2554162	3263582	"Pzl"	709420
3R	2744304	2744800	"CR45220"	496
3R	3322810	3354486	"spok"	31676
3R	3461351	3587054	"Parp"	125703


d. Instead of printing out first 10 lines, now redirect the results to a new file named `BDGP6.Ensembl.93.gene.bed`. Verify that your file contains 17721 lines.


In [ ]:
!awk '$3 == "gene" {print $1 "\t" $4 "\t" $5 "\t" $12 "\t" $5-$4}' BDGP6.Ensembl.93.gtf \
| grep -v ^211 | sed s/';'//g > BDGP6.Ensembl.93.gene.bed

In [ ]:
# verify the number of lines

!wc -l BDGP6.Ensembl.93.gene.bed

17721 BDGP6.Ensembl.93.gene.bed


e. Save the first 100 lines of the results to `~/5030/hw1/q3/output/BDGP6.Ensembl.93.head100.bed`


In [ ]:
!awk '$3 == "gene" {print $1 "\t" $4 "\t" $5 "\t" $12 "\t" $5-$4}' BDGP6.Ensembl.93.gtf \
| grep -v ^211 | sed s/';'//g | head -100 > ../output/BDGP6.Ensembl.93.head100.bed

In [ ]:
# verify the number of lines

!wc -l ../output/BDGP6.Ensembl.93.head100.bed

100 ../output/BDGP6.Ensembl.93.head100.bed


### 5. Count number of genes per chromosome in `BDGP6.Ensembl.93.gene.bed` using `cut` `sort` and `uniq`. Redirect your results to `~/5030/hw1/q3/output/genes_per_chromosome.out`.


In [ ]:
# visualizing the data

!head BDGP6.Ensembl.93.gene.bed

3R	567076	2532932	"Myo81F"	1965856
3R	722370	722621	"CR41571"	251
3R	1031171	1031354	"CR12798"	183
3R	1366234	1366601	"CR46123"	367
3R	1865108	1866008	"CR46122"	900
3R	2156916	2157206	"CR40182"	290
3R	2554162	3263582	"Pzl"	709420
3R	2744304	2744800	"CR45220"	496
3R	3322810	3354486	"spok"	31676
3R	3461351	3587054	"Parp"	125703


In [ ]:
# redirecting the results

!cut -f1 BDGP6.Ensembl.93.gene.bed | sort | uniq -c | sort -nr > ../output/genes_per_chromosome.out

In [ ]:
# visualizing the results

! head ../output/genes_per_chromosome.out

   4191 3R
   3621 2R
   3496 2L
   3457 3L
   2671 X
    113 Y
    111 4
     38 mitochondrion_genome
     21 rDNA
      2 Unmapped_Scaffold_8


### 6. Find 20 longest and shortest genes in `BDGP6.Ensembl.93.gene.bed` using `sort` command with `-n` and `-k` options. Redirect the results in `~/5030/hw1/q3/output/longest_20_genes.out` and `~/5030/hw1/q3/output/shortest_20_genes.out` respectively. (Hint: First of all, figure out what are options `-n` and `-k` in the `sort` command. Then take advantage of the commands that allow you to examine the beginning or the end of a file.) Your output file should look like


  ```
  3R      31846987        32015520        "heph"  168534
  X       5085759 5254864 "CG15465"       169106
  ```

In [ ]:
# the longest 20 genes

!sort -k5 -nr BDGP6.Ensembl.93.gene.bed | head -20 > ../output/longest_20_genes.out

In [ ]:
# the shortest 20 genes

!sort -k5 -n BDGP6.Ensembl.93.gene.bed | head > ../output/shortest_20_genes.out

In [ ]:
# visualizing the results (long)

!head ../output/longest_20_genes.out

3R	567076	2532932	"Myo81F"	1965856
3R	2554162	3263582	"Pzl"	709420
3L	25236940	25670980	"CG45782"	434040
2R	1866080	2262115	"CG17684"	396035
X	2740384	3134532	"kirre"	394148
2R	326768	705848	"CG45781"	379080
X	12043698	12335214	"Ten-a"	291516
3L	13521907	13803400	"bru3"	281493
Y	1636574	1884846	"Ppr-Y"	248272
X	7166451	7405797	"CR44357"	239346


In [ ]:
# visualising the results (short)

!head ../output/shortest_20_genes.out

rDNA	16526	16555	"2SrRNA-Psi:CR45850"	29
rDNA	46356	46385	"2SrRNA:CR45836"	29
rDNA	58805	58834	"2SrRNA:CR45840"	29
rDNA	70540	70569	"2SrRNA:CR45843"	29
X	23205409	23205438	"2SrRNA-Psi:CR40677"	29
X	23214040	23214069	"2SrRNA-Psi:CR45858"	29
X	23291700	23291729	"2SrRNA:CR45864"	29
2L	21680670	21680700	"snmRNA:254"	30
2L	19403535	19403568	"snmRNA:359"	33
2R	21335510	21335544	"snmRNA:184"	34


7. Use `awk` command to find genes that are longer than 5000 bps in `BDGP6.Ensembl.93.gene.bed` and redirect the results to `~/5030/hw1/q3/output/BDGP6.Ensembml.93.gt5000.bed`.


In [ ]:
!awk '$5 > 5000 {print $0}' BDGP6.Ensembl.93.gene.bed > ../output/BDGP6.Ensembml.93.gt5000.bed

In [ ]:
# visualizing the output

!head ../output/BDGP6.Ensembml.93.gt5000.bed

3R	567076	2532932	"Myo81F"	1965856
3R	2554162	3263582	"Pzl"	709420
3R	3322810	3354486	"spok"	31676
3R	3461351	3587054	"Parp"	125703
3R	3734855	3745306	"Tim17b"	10451
3R	3846729	3868401	"CG41099"	21672
3R	4077939	4104326	"CG42402"	26387
3R	4126442	4137882	"Gfat1"	11440
3R	4174632	4186751	"CG12581"	12119
3R	4186963	4196210	"CR33294"	9247


In [ ]:
# verifying that the min is not smaller than 5000

!sort ../output/BDGP6.Ensembml.93.gt5000.bed | head -3

2L	10003832	10008889	"CG4747"	5057
2L	10032690	10038954	"obst-B"	6264
2L	10064121	10180256	"CG44153"	116135


### 8. Starting again with the original `.gtf` file. Find the top 10 genes with most transcripts using `grep` `sort` and `tail` (or `head`).

In [ ]:
# visualizing the data

!head BDGP6.Ensembl.93.gtf

#!genome-build BDGP6
#!genome-version BDGP6
#!genome-date 2014-07
#!genome-build-accession GCA_000001215.4
3R	FlyBase	gene	567076	2532932	.	+	.	gene_id "FBgn0267431"; gene_name "Myo81F"; gene_source "FlyBase"; gene_biotype "protein_coding";
3R	FlyBase	transcript	567076	2532932	.	+	.	gene_id "FBgn0267431"; transcript_id "FBtr0392909"; gene_name "Myo81F"; gene_source "FlyBase"; gene_biotype "protein_coding"; transcript_name "Myo81F-RB"; transcript_source "FlyBase"; transcript_biotype "protein_coding";
3R	FlyBase	exon	567076	567268	.	+	.	gene_id "FBgn0267431"; transcript_id "FBtr0392909"; exon_number "1"; gene_name "Myo81F"; gene_source "FlyBase"; gene_biotype "protein_coding"; transcript_name "Myo81F-RB"; transcript_source "FlyBase"; transcript_biotype "protein_coding"; exon_id "FBtr0392909-E1";
3R	FlyBase	exon	835376	835491	.	+	.	gene_id "FBgn0267431"; transcript_id "FBtr0392909"; exon_number "2"; gene_name "Myo81F"; gene_source "FlyBase"; gene_biotype "protein_coding"; transcript_name 

  a. Use `grep -w` to get rows of transcript features (column 3). There should be a total of 34767 lines.


In [ ]:
!cut -f3 BDGP6.Ensembl.93.gtf | grep -v "^#" | grep -w transcript | wc -l

34767


  b. Then use `cut` `sort` `uniq -c` and `sort` again to order the genes by their occurrences. Redirect the top 10 genes to `~/5030/hw1/q3/output/genes_w_most_transcripts.out`. Perform the task in a single line of command. Then display contents in `genes_w_most_transcripts.out` using `cat` command. They should look like


  ```
  23 "slo";
  25 "heph";
  25 "lola";
  ```

In [ ]:
!awk '$3 == "transcript" {print $0}' BDGP6.Ensembl.93.gtf | cut -f9 | cut -d'"' -f6 \
| sort | uniq -c | sort -nr | head -10

     75 Dscam1
     60 para
     31 Svil
     25 Sxl
     25 lola
     25 heph
     23 slo
     23 mtd
     23 CAP
     22 trol


In [ ]:
# another method (to keep the '"')

!awk '$3 == "transcript" {print $0}' BDGP6.Ensembl.93.gtf | cut -f9 | awk '{print $6}' | sed s/';'//g \
| sort | uniq -c | sort -nr | head -10

     75 "Dscam1"
     60 "para"
     31 "Svil"
     25 "Sxl"
     25 "lola"
     25 "heph"
     23 "slo"
     23 "mtd"
     23 "CAP"
     22 "trol"


In [ ]:
# saving the output

!awk '$3 == "transcript" {print $0}' BDGP6.Ensembl.93.gtf | cut -f9 | awk '{print $6}' | sed s/';'//g | sort | uniq -c | sort -nr | head -10 > ../output/genes_w_most_transcripts.out

### 9. Submit the output files.

a. Summarize your directory structure again by saving the output of `ls -lR ~/5030/hw1/q3` into the `summary` directory as a file named `q9_directory_contents.txt`.


In [ ]:
%cd ..

/content/5030/hw1/q3


In [ ]:
!ls -lR . > summary/q9_directory_contents.txt

In [ ]:
# visualizing the output

!head -30 summary/q9_directory_contents.txt

.:
total 12
drwxr-xr-x 2 root root 4096 Feb  2 13:47 genome-annot
drwxr-xr-x 2 root root 4096 Feb  2 14:32 output
drwxr-xr-x 2 root root 4096 Feb  2 14:34 summary

./genome-annot:
total 600
-rw-r--r-- 1 root root 610013 Feb  2 13:47 BDGP6.Ensembl.93.gene.bed
lrwxrwxrwx 1 root root     72 Feb  2 08:26 BDGP6.Ensembl.93.gtf -> /content/drive/MyDrive/Colab Notebooks/5030/genomes/BDGP6.Ensembl.93.gtf

./output:
total 140
-rw-r--r-- 1 root root   3287 Feb  2 13:49 BDGP6.Ensembl.93.head100.bed
-rw-r--r-- 1 root root 121422 Feb  2 14:14 BDGP6.Ensembml.93.gt5000.bed
-rw-r--r-- 1 root root    144 Feb  2 14:01 genes_per_chromosome.out
-rw-r--r-- 1 root root    148 Feb  2 14:32 genes_w_most_transcripts.out
-rw-r--r-- 1 root root    684 Feb  2 14:06 longest_20_genes.out
-rw-r--r-- 1 root root    391 Feb  2 14:06 shortest_20_genes.out

./summary:
total 4
-rw-r--r-- 1 root root 427 Feb  2 08:28 q3_directory_contents.txt
-rw-r--r-- 1 root root   0 Feb  2 14:34 q9_directory_contents.txt


b. Use `tar -czvf` to create a `.tar.gz` file named `hw1_q3.tar.gz`, containing your directories of `~/5030/hw1/q3/output` and `~/5030/hw1/q3/summary`. Then copy `hw1_q3.tar.gz` to your folder under `/hdd1/5030-hw`.


In [ ]:
!tar -czvf hw1_q3.tar.gz output summary

output/
output/BDGP6.Ensembml.93.gt5000.bed
output/BDGP6.Ensembl.93.head100.bed
output/genes_w_most_transcripts.out
output/genes_per_chromosome.out
output/shortest_20_genes.out
output/longest_20_genes.out
summary/
summary/q3_directory_contents.txt
summary/q9_directory_contents.txt


the file hw1_q3.tar.gz was downloaded separately as this homework was done in the google colab environment

In [ ]:
# !cp hw1_q3.tar.gz ../../../hdd1/5030-hw

### 10. Clean up.

  a. Use `find` command to find all `.bed` files that are larger than 50kb under `~/5030/hw1/q3`.

In [ ]:
!find . -name '*.bed' -size +50b

./genome-annot/BDGP6.Ensembl.93.gene.bed
./output/BDGP6.Ensembml.93.gt5000.bed


 b. Modify the command, e.g. using the backquote sign (\`), to find and remove any file that is larger than 500kb under `~/5030/hw1/q3`.


In [ ]:
!rm `find . -name '*.bed' -size +50b`

In [ ]:
# checking out
!find . -name '*.bed' -size +50b